Feature: Opus as additional audio codec

susnux opened the pull request:
I was asked to create this PR, so this will add opus as audio codec.
As described in the [ Forum thread](http://en.sfml-dev.org/forums/index.php?topic=19856.0).

If you prefer one commit I can squash them into one.
Feel free to add comments.

LaurentGomila commented:
Looks very good. You should just put empty lines above comment lines, and I'd also rewrite the `writeint` macro as a function, like it's done in other encoders.

Is error.opus small enough, free of charge and under a suitable license?

I think you've forgotten to add the opus headers and libraries (in /extlibs).

Eremiell commented:
Opus is a next gen xiph.org codec [licensed under 3-clause BSD](http://opus-codec.org/license/) that should [according to tests beat both Speex and Ogg Vorbis in their respective fields](http://opus-codec.org/comparison/).

For size

| file | size | file | size | file | size | file | size |
| --- | --- | --- | --- | --- | --- | --- | --- |
| libvorbis.so | 175k | libvorbis.a | 233k | libvorbisfile.so | 35k | libvorbisfile.a | 34k |
| libopus.so | 294k | libopus.a | 561k | libopusfile.so | 46k | libopusfile.a | 54k |

on my x86_64 Debian machine, sizes on other platforms should be similar. It is slightly bigger than Vorbis, but not much and covers both speech and music.

susnux commented:
@LaurentGomila 
Sample file: It is public domain through the CC 0 license, took it from this place [Freesound](http://freesound.org/people/Ziembee/sounds/161694/). But I will replace it with an other encode, I noticed it is unneeded big.

Also with the last commit I think I implemented your notes (empty lines, rewrote writeint).

susnux commented:
Ahh the static android libraries are missing, will commit them.
I had to modify the build script because the new NDK uses gcc toolchain 4.9 not 4.8 thats why it failed for me...

LaurentGomila commented:
The MSVC libraries can be built by someone else, don't worry.

The `writeInt` function still needs some formatting for compliance with our coding rules. I adapted a similar function found in the WAV encoder, for consistency. I also removed the `offset` argument, since it can simply be added to `buffer` when the function is called.

```cpp
    // Make sure to write int into buffer little endian
    void writeInt(unsigned char* buffer, sf::Uint32 value)
    {
            buffer[0] = static_cast<unsigned char>(value & 0x000000FF),
            buffer[1] = static_cast<unsigned char>((value & 0x0000FF00) >> 8),
            buffer[2] = static_cast<unsigned char>((value & 0x00FF0000) >> 16),
            buffer[3] = static_cast<unsigned char>((value & 0xFF000000) >> 24),
    }
```

susnux commented:
Ok changed that, removed include it is useless (I used it to debug but forgot to remove it).

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+bool SoundFileWriterOpus::check(const std::string& filename)
+{
+    std::string extension = filename.substr(filename.find_last_of(".") + 1);
+    std::transform(extension.begin(), extension.end(), extension.begin(), ::tolower);
```
`tolower()` is -- like everything in the C++ standard library -- part of the `std` namespace. The function also requires an additional `#include <cctype>`. See [here](http://en.cppreference.com/w/cpp/string/byte/tolower) for more information.

Bromeon replied:
Adding answer to [Laurent's comment](https://github.com/SFML/SFML/pull/1077#issuecomment-204992310) here:

> He just copied the code that I wrote in other encoders. And I probably did that on purpose (I never explicitly scope a symbol just for fun). Maybe the C++  tolower  doesn't compile here due to template/overload stuff? I don't remember.

There's an additional overload of `std::tolower` in `<locale>` which takes a locale as second parameter. When this header is included, there may indeed be an ambiguity. So I guess you encountered problems because some standard libraries include that header through others... In that case you're right, the `::tolower` might be better.

...or to ensure 100% portability, `static_cast<int (*)(int)>(&std::tolower)` :stuck_out_tongue:

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+        close();
+        return false;
+    }
+    int status = 0;
```
Is there no constant to denote an invalid status value? I'm asking because it may be possible that `OPUS_OK` can be defined as 0. Also, it would make code more clear (`0` is a magic constant here).

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+    {
+        err() << "Failed to write ogg/opus file \"" << filename << "\"" << std::endl;
+        if (status == OPUS_BAD_ARG)
+            err() << "Possible wrong sample rate, allowed are 8000, 12000, 16000, 24000, or 48000 Hz." << std::endl;
```
I would replace "Possible wrong" with "Invalid" or similar

susnux replied:
@Bromeon Hm but it might be something different then the sample rate. The sample rate is only the most common issue here. E.g. the default sfml sample rate is 44,1kHz which is not supported, if really needed I could add a function to interpolate the sample rate to the next supported, but I think this would be overkill for sfml's needs, or?

Bromeon replied:
Then you mean "Possi**bly** wrong sample rate"?

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+    unsigned char commentData[commentLength];
+
+    // Magic bytes
+    memcpy(static_cast<void*>(commentData), "OpusTags", 8);
```
`std::memcpy()`

susnux replied:
> std::memcpy -- and conversion to void* can be implicit.
Also in other places...

Ok I will add the namespace, my intention for the cast was to avoid compiler warnings (-Wpedantic).

Bromeon replied:
Does an implicit conversion from T* to void* really yield a warning? It's so annoying when compilers emit warnings for perfectly valid, justified and in no way dangerous C++ code.

I thought the `-pedantic` was intended to enforce strict conformance to the C++ standard. It's pretty ironic that the variable-length array is *not* reported.

LaurentGomila commented:
> tolower() is -- like everything in the C++ standard library -- part of the std namespace

He just copied the code that I wrote in other encoders. And I probably did that on purpose (I never explicitly scope a symbol just for fun). Maybe the C++ version don't compile here due to template/overload stuff? I don't remember.

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+    Uint32 opusVersionLength = strlen(opusVersion);
+    const int commentLength  = 8 + 4 + opusVersionLength + 4;
+
+    unsigned char commentData[commentLength];
```
Not allowed in C++, because `commentLength` is not a constant expression. Use `std::vector` instead.

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+    memcpy(static_cast<void*>(commentData+12), opusVersion, opusVersionLength);
+
+    // Length of user comments (E.g. you can add a ENCODER tag for SFML)
+    writeUint32(commentData + 12+opusVersionLength, 0);
```
Spaces (also above)

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+    unsigned char buffer[buffer_size];
+    Uint32 frame_number = 0;
+    Uint8 endOfStream = 0;
+    ogg_packet op;
```
I would declare the variables that are not reused between iterations (such as `op`) locally inside the loop. It makes code easier to understand.

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+        opus_int32 packet_size;
+
+        // Check if wee need to pad the input
+        if (count < (frame_size * m_channelCount))
```
Unneeded parentheses

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+        {
+            Int16 pad[frame_size*m_channelCount];
+            memset(pad, 0, frame_size * m_channelCount);
+            memcpy(pad, samples + (frame_number * frame_size * m_channelCount), count);
```
Unneeded parentheses

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+void SoundFileWriterOpus::flushBlocks()
+{
+    ogg_page page;
+    while (ogg_stream_pageout(&m_ogg, &page) > 0 || ogg_stream_flush(&m_ogg, &page) > 0)
```
Here, we add parentheses around each `||` operand.

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    // Clear all the ogg/opus structures
+    ogg_stream_clear(&m_ogg);
```
Will this crash if `close()` is called twice?

I'm asking because the other variables (`m_file` and `m_opus`) are checked for being deallocated already, but not this one...

susnux replied:
> Will this crash if close() is called twice?

No it is safe:

> It is safe to call ogg_stream_clear on the same structure more than once.

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+        // Check if wee need to pad the input
+        if (count < (frame_size * m_channelCount))
+        {
+            Int16 pad[frame_size*m_channelCount];
```
Not valid C++

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+
+    const opus_uint32 frame_size = 960;
+    const opus_int32 buffer_size = frame_size * m_channelCount * sizeof(Int16);
+    unsigned char buffer[buffer_size];
```
Not valid C++

Bromeon commented on src/SFML/Audio/SoundFileReaderOpus.cpp:
```diff
@@ -0,0 +1,194 @@
+{
+    if (m_opus != NULL)
+    {
+        op_free(m_opus);
```
Shouldn't `m_opus` be set to `NULL` here?

Bromeon commented on src/SFML/Audio/SoundFileReaderOpus.cpp:
```diff
@@ -0,0 +1,194 @@
+    Uint64 count = 0;
+    while (maxCount > 0)
+    {
+
```
Unneeded newline

Bromeon commented on src/SFML/Audio/SoundFileReaderOpus.cpp:
```diff
@@ -0,0 +1,194 @@
+        return static_cast<opus_int64>(stream->tell());
+    }
+
+    static OpusFileCallbacks callbacks = {&read, &seek, &tell, NULL};
```
`static` is not needed; `callbacks` is defined in an anonymous namespace and thus already has internal linkage.

I'd consider making it `const` however.

Bromeon commented:
Thanks a lot for your contribution! The code looks mostly good, I've added some comments (mostly related to style or non-standard C++).

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,259 @@
+    // Save the channel count
+    m_channelCount = channelCount;
+    m_sampleRate = sampleRate;
+    std::srand(std::time(0));
```
It's rather intrusive to seed the global random generator. As a user I may have an application that depends on deterministic behavior of `rand()`, and I'd be very annoyed if a library messes with that without being told so...

susnux commented:
sfml is not going to use c++11 or is it?
If not I have to rewrite my changes (used std::vector with c++11 only functions to replace non c++ dynamic arrays (c99)). Like initializer lists and std::vector:data()

Bromeon commented:
SFML currently uses C++98. Which doesn't matter, because variable-length arrays are not part of any C++ standard. They have been introduced to C99, but never to C++.

As an alternative, a library-based solution was discussed for C++14, namely `std::dynarray`, however it was voted out [[1](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2008/n2648.html), [2](http://en.cppreference.com/w/cpp/container/dynarray)].

susnux commented:
@Bromeon Thank you!
I think I resolved all your comments.

> SFML currently uses C++98, and it will use C++11 starting from a future major version 3.

Ok I replaced the dynamic arrays with std::vector where only read support is needed and the other dynamic buffer with heap memory (new).
With c++11 some lines can be improved (&vector.front() can be replaced with vector.data(), initalizer_list constructors...), but I think it is ok for the moment.

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,254 @@
+        // Check if wee need to pad the input
+        if (count < frame_size * m_channelCount)
+        {
+            const Uint32 begin = (frame_number * frame_size * m_channelCount);
```
Unneeded parentheses (also in several other snippets). Generally, use operator precedence to your advantage, and add parentheses only where it actually clarifies code or where the code style demands it (for `||` and `&&` operators).

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,254 @@
+        return false;
+    }
+
+    int status = OPUS_OK;
```
What I meant before is rather to initialize the status to something that's not OK. I don't know about `opus_encoder_create`'s semantics, but there's a chance that it does not fill the status flag for some errors.

My thought was simply:
"by default, loading is not successful, unless the loading function explicitly says so"
rather than
"loading is successful, unless the loading function says otherwise" :grinning:

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,254 @@
+
+    const opus_uint32 frame_size = 960;
+    const opus_int32 buffer_size = frame_size * m_channelCount * sizeof(Int16);
+    unsigned char* buffer = new unsigned char[buffer_size];
```
use `std::vector<unsigned char> buffer(bufferSize);`

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,254 @@
+{
+    assert(m_opus);
+
+    const opus_uint32 frame_size = 960;
```
Naming convention: we don't use underlines (expect for macros). Use `frameSize`.

Bromeon commented on src/SFML/Audio/SoundFileWriterOpus.cpp:
```diff
@@ -0,0 +1,254 @@
+    while (count > 0)
+    {
+        ogg_packet op;
+        opus_int32 packet_size;
```
Those variables are not used in the following code, you should declare them as late as possible. `packet_size` should be `packetSize`.

Bromeon commented:
Looks much better, thanks for adapting! There are still some notes, sorry for being picky :grimacing:
Maybe you should have a look at the [style guide](http://www.sfml-dev.org/style.php). For minor things, I guess we could also do a final touch, instead of annotating every single misplaced character.

> Ok I replaced the dynamic arrays with std::vector where only read support is needed and the other dynamic buffer with heap memory (new).

Avoid `new[]` and use `std::vector` everywhere. You can write to vectors just as easily. Dynamic memory management is generally a bad idea, I wrote [an article](http://www.bromeon.ch/articles/raii.html) about it in case you're interested.

> With c++11 some lines can be improved (&vector.front() can be replaced with vector.data(), initializer_list constructors...), but I think it is ok for the moment.

Yeah, that looks good. Your comments regarding C++11 migration are a nice help!

susnux commented:
No problem ;-)

Ok changed  that.
Thank you for the link, I will read it later. I personally have no problem with new and delete in a function, but I dislike using it between functions / classes when the owner is not clear.

CoolOppo commented:
How is this coming along? I would absolutely love to have Opus support in SFML!

eXpl0it3r commented:
@susnux Could you rebase this on to master?

susnux commented:
@eXpl0it3r ok rebased the branch.

LaurentGomila commented:
Minor suggestion: I would not add something to the example, especially if it adds a new resource without demonstrating something new in the API (the goal is to keep SDK size as low as possible).

susnux commented:
@LaurentGomila I have added the example, because ogg and flac also added an example for playing a flac/ogg file, I can remove it if you prefer that.

eXpl0it3r commented:
Yeah, I added the FLAC file to the example as well and I think it's okay to add ome for Opus as well, as long as it's reasonably small.

MarioLiebisch commented:
I agree with Laurent. Wouldn't add anything, since it's really just another file format used the very same way with nothing new to show. It's just a different string passed as a parameter. If you spin that further we'd have to add Windows Bitmaps and other image formats as well.

mantognini commented:
I don't mind another small resource like this one personally. It's very convenient to have those examples around when we want to quickly test a feature/bugfix. In any case, we might want to move all examples to a dedicated git repo at some point.

LaurentGomila commented:
Yes, I agree that it can be convenient for quick testing after modifications to the audio module. But for that we'd better have proper (unit) tests ;)

mantognini commented:
Definitely, but at the same time we are so far away from a usable unit test suite that this seems more than reasonable to me.

LaurentGomila commented:
Alright.

eXpl0it3r commented:
I'll look into building Windows libraries.

eXpl0it3r commented:
So I've been trying to build opus for Windows for the better part of the day and kept running into linker issues. Talking with someone in the opus IRC channel, it became clear, that for opus below version 1.2 (currently this PR uses 1.1.2), one not only has to link opus and opusfile, but also celt and silk. Version 1.2 which currently is in alpha combines celt and silk directly into opus.

The FindOpus.cmake script doesn't seem to know about these celt or silk libraries as such I'm curious how you managed to build SFML on your system?

Finally I'm not sure whether we really want to add four new dependencies.

LaurentGomila commented:
If dependencies become an issue, we can maybe think about adding some codecs as plugins. The classes are already made for such usage, we'd just have to add a plugin loader/manager to the audio module.

susnux commented:
@eXpl0it3r 
> I'm curious how you managed to build SFML on your system?

I just installed the libraries required by SFML and in addition libopus and libopusfile.
I am developing and testing on openSUSE Leap 42.1, so I have libopus 1.1 and libopusfile 0.8.
SFML builds successfully, no linking issue and it seems like there is no special magic when building libopus:

```./configure --disable-silent-rules  --disable-doc --enable-custom-modes```

Here is my build log of SFML with opus support: [log2.txt](https://github.com/SFML/SFML/files/586228/log2.txt)

There is also no celt in linked libraries of sfml-audio.so:
```
% ldd lib/libsfml-audio.so
        linux-vdso.so.1 (0x00007ffc99c05000)
        libsfml-system.so.2.4 => /home/ferdinand/Dokumente/Projekte/C++/SFML/build/lib/libsfml-system.so.2.4 (0x00007f2a503eb000)
        libopenal.so.1 => /usr/lib64/libopenal.so.1 (0x00007f2a50143000)
        libvorbisenc.so.2 => /usr/lib64/libvorbisenc.so.2 (0x00007f2a4fc73000)
        libvorbisfile.so.3 => /usr/lib64/libvorbisfile.so.3 (0x00007f2a4fa63000)
        libvorbis.so.0 => /usr/lib64/libvorbis.so.0 (0x00007f2a4f833000)
        libogg.so.0 => /usr/lib64/libogg.so.0 (0x00007f2a4f62b000)
        libopusfile.so.0 => /usr/lib64/libopusfile.so.0 (0x00007f2a4f41b000)
        libopus.so.0 => /usr/lib64/libopus.so.0 (0x00007f2a4f1cb000)
        libFLAC.so.8 => /usr/lib64/libFLAC.so.8 (0x00007f2a4ef8b000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f2a4ed6b000)
        librt.so.1 => /lib64/librt.so.1 (0x00007f2a4eb63000)
        libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f2a4e7db000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f2a4e4d3000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f2a4e2bb000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f2a4df13000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f2a4dd0b000)
        libcrypto.so.1.0.0 => /lib64/libcrypto.so.1.0.0 (0x00007f2a4d913000)
        /lib64/ld-linux-x86-64.so.2 (0x0000562c8d37e000)
        libz.so.1 => /lib64/libz.so.1 (0x00007f2a4d6fb000)
```

eXpl0it3r commented:
I talked with the same person again on IRC and it seems I misunderstood them. The split of these libraries only exists in the VS solution and was merged after 1.1.3 into one VS project.

I'll try and see if one can just move all the files from the other projects into one a build it like that.

eXpl0it3r commented:
I have picked this up again. After spending a lot of hours getting all the libraries to build, the MinGW version finally builds and links just fine, unfortunately the example opus file doesn't play a sound. Next I'll have to build opusfile for VS and see if it works there. I'll certainly keep investigating this.

Example code:

```cpp
#include <SFML/Audio.hpp>
#include <iostream>

int main()
{
    sf::Music music;
    if (!music.openFromFile("file.opus"))
        return 0;

    std::cout << " " << music.getDuration().asSeconds() << " seconds"       << std::endl;
    std::cout << " " << music.getSampleRate()           << " samples / sec" << std::endl;
    std::cout << " " << music.getChannelCount()         << " channels"      << std::endl;

    music.play();

    while (music.getStatus() == sf::Music::Playing)
    {
        sf::sleep(sf::milliseconds(100));

        std::cout << "\rPlaying... " << music.getPlayingOffset().asSeconds() << " sec        ";
        std::cout << std::flush;
    }
    std::cout << std::endl << std::endl;

    std::cout << "Press enter to exit..." << std::endl;
    std::cin.ignore(10000, '\n');
}
```

eXpl0it3r commented:
This PR has been superseded by #1259!

Let me know if there are some additional changes to re-sync.
